File: go/internal/sol/p0035/p0035.go
package p0035

import (
	"strconv"

	"pe-solver/internal/mylib"
)

func isCircularPrime(n int) bool {
	// 'n' must be a prime number.
	k := mylib.NumOfDigits(n, 10) - 1
	d := mylib.Pow(10, k)

	for k > 0 {
		if n = (n%10)*d + n/10; !mylib.IsPrime(n) {
			return false
		}
		k--
	}

	return true
}

func compute(limit int) string {
	var result int
	for _, prime := range mylib.Primes(limit) {
		if isCircularPrime(prime) {
			result++
		}
	}

	return strconv.FormatInt(int64(result), 10)
}

func Solve() string {
	return compute(1_000_000)
}

File: haskell/src/Sol/P0035.hs
module Sol.P0035 (compute, solve) where

import Mylib.Math (numOfDigits)
import Mylib.Prime (isPrime, primes)

compute :: Int -> String
compute limit =
    show $ length $ filter isCircularPrime $ primes 1 limit
  where
    isCircularPrime :: Int -> Bool
    isCircularPrime n =
        all
            isPrime
            (drop 1 . take (k + 1) $ iterate (\x -> (mod x 10) * d + (div x 10)) n)
      where
        k = (numOfDigits n 10) - 1
        d = 10 ^ k

solve :: String
solve = compute 1_000_000
